Plutarch/Api/V1.hs

```diff
@@ -0,0 +1,658 @@
+
+newtype PTxOut (s :: S)
+  = PTxOut
+      ( Term
+          s
+          ( PDataList
+              '[ PAddress
+               , PValue
+               , PMaybe PDatumHash
+               ]
+          )
+      )
+  deriving
+    (PMatch, PIsData)
+    via (PIsDataReprInstances PTxOut)
+
+instance PIsDataRepr PTxOut where
+  type
+    PIsDataReprRepr PTxOut =
+      '[ '[ PAddress
+          , PValue
+          , PMaybe PDatumHash
+          ]
+       ]
+
+  pmatchRepr dat f =
+    pmatchDataRepr dat $
+      DRHCons (f . PTxOut) DRHNil
+
+data PDCert (s :: S)
+  = PDCertDelegRegKey (Term s (PDataList '[PStakingCredential]))
+  | PDCertDelegDeRegKey (Term s (PDataList '[PStakingCredential]))
+  | PDCertDelegDelegate (Term s (PDataList '[PStakingCredential, PPubKeyHash]))
+  | PDCertPoolRegister (Term s (PDataList '[PPubKeyHash, PPubKeyHash]))
+  | PDCertPoolRetire (Term s (PDataList '[PPubKeyHash, PInteger]))
+  | PDCertGenesis (Term s (PDataList '[]))
+  | PDCertMir (Term s (PDataList '[]))
+  deriving
+    (PMatch, PIsData)
+    via (PIsDataReprInstances PDCert)
+
+instance PIsDataRepr PDCert where
+  type
+    PIsDataReprRepr PDCert =
+      '[ '[PStakingCredential]
+       , '[PStakingCredential]
+       , '[PStakingCredential, PPubKeyHash]
+       , '[PPubKeyHash, PPubKeyHash]
+       , '[PPubKeyHash, PInteger]
+       , '[]
+       , '[]
+       ]
+
+  pmatchRepr dat f =
+    pmatchDataRepr dat $
+      DRHCons (f . PDCertDelegRegKey) $
+        DRHCons (f . PDCertDelegDeRegKey) $
+          DRHCons (f . PDCertDelegDelegate) $
+            DRHCons (f . PDCertPoolRegister) $
+              DRHCons (f . PDCertPoolRetire) $
+                DRHCons (f . PDCertGenesis) $
+                  DRHCons (f . PDCertMir) DRHNil
+
+---------- AssocMap
+
+newtype PMap (k :: PType) (v :: PType) (s :: S) = PMap (Term s (PBuiltinMap k v))
+  deriving (PlutusType, PIsData) via (DerivePNewtype (PMap k v) (PBuiltinMap k v))
+
+instance
+  ( Plutus.ToData (PLifted v)
+  , Plutus.ToData (PLifted k)
+  , Plutus.FromData (PLifted v)
+  , Plutus.FromData (PLifted k)
+  , PLift k
+  , PLift v
+  ) =>
+  PUnsafeLiftDecl (PMap k v)
+  where
+  type PLifted (PMap k v) = PlutusMap.Map (PLifted k) (PLifted v)
+
+instance
+  ( PLifted (PConstanted k) ~ k
+  , Plutus.ToData v
+  , Plutus.FromData v
+  , Plutus.ToData k
+  , Plutus.FromData k
+  , PConstant k
+  , PLifted (PConstanted v) ~ v
+  , Plutus.FromData v
+  , Plutus.ToData v
+  , PConstant v
+  ) =>
+  PConstant (PlutusMap.Map k v)
+  where
+  type PConstantRepr (PlutusMap.Map k v) = [(Plutus.Data, Plutus.Data)]
+  type PConstanted (PlutusMap.Map k v) = PMap (PConstanted k) (PConstanted v)
+  pconstantToRepr m = (\(x, y) -> (Plutus.toData x, Plutus.toData y)) <$> PlutusMap.toList m
+  pconstantFromRepr m = fmap PlutusMap.fromList $
+    flip traverse m $ \(x, y) -> do
+      x' <- Plutus.fromData x
+      y' <- Plutus.fromData y
+      Just (x', y')
+
+---------- Others
+
+data PMaybe a (s :: S)
```

Comment:
These names conflict with those of the scott encoded versions. Have we got a policy for this yet? Perhaps these should be moved out to a separate module or explicitly renamed? It's a bit of a pain to import both `Plutarch.Api.V1` and `Plutarch.Maybe` at the same time during dev.

Reply:
That's also what I'm thinking.

Reply:
We need to revisit this.

Reply:
cf. https://github.com/Plutonomicon/plutarch/issues/106#issuecomment-1012103112 

and https://github.com/Plutonomicon/plutarch/issues/106#issuecomment-1013019797